File: goldenflow/transforms/numeric.py
from __future__ import annotations

import re

import polars as pl

from goldenflow.transforms import register_transform


@register_transform(
    name="currency_strip", input_types=["string", "numeric"], auto_apply=False, priority=50, mode="series"
)
def currency_strip(series: pl.Series) -> pl.Series:
    def _strip(val: str | None) -> float | None:
        if val is None:
            return None
        cleaned = re.sub(r"[^\d.\-]", "", str(val))
        try:
            return float(cleaned)
        except ValueError:
            return None

    return series.map_elements(_strip, return_dtype=pl.Float64)


@register_transform(
    name="percentage_normalize",
    input_types=["string", "numeric"],
    auto_apply=False,
    priority=50,
    mode="series",
)
def percentage_normalize(series: pl.Series) -> pl.Series:
    def _norm(val: str | None) -> float | None:
        if val is None:
            return None
        v = str(val).strip().rstrip("%")
        try:
            return float(v) / 100.0
        except ValueError:
            return None

    return series.map_elements(_norm, return_dtype=pl.Float64)


@register_transform(
    name="round", input_types=["numeric"], auto_apply=False, priority=40, mode="series"
)
def round_values(series: pl.Series, n: int = 2) -> pl.Series:
    return series.round(n)


@register_transform(
    name="clamp", input_types=["numeric"], auto_apply=False, priority=40, mode="series"
)
def clamp(series: pl.Series, min_val: float = 0.0, max_val: float = 1.0) -> pl.Series:
    return series.clip(min_val, max_val)


@register_transform(
    name="to_integer",
    input_types=["string", "numeric"],
    auto_apply=False,
    priority=45,
    mode="series",
)
def to_integer(series: pl.Series) -> pl.Series:
    """Parse string to integer, truncating any decimal part."""

    def _to_int(val: str | None) -> int | None:
        if val is None:
            return None
        try:
            return int(float(val))
        except (ValueError, OverflowError):
            return None

    return series.map_elements(_to_int, return_dtype=pl.Int64)


@register_transform(
    name="abs_value",
    input_types=["numeric"],
    auto_apply=False,
    priority=40,
    mode="series",
)
def abs_value(series: pl.Series) -> pl.Series:
    """Return the absolute value."""
    return series.abs()


@register_transform(
    name="fill_zero",
    input_types=["numeric"],
    auto_apply=False,
    priority=35,
    mode="series",
)
def fill_zero(series: pl.Series) -> pl.Series:
    """Replace null values with 0."""
    return series.fill_null(0)


@register_transform(
    name="comma_decimal",
    input_types=["string", "numeric"],
    auto_apply=False,
    priority=48,
    mode="series",
)
def comma_decimal(series: pl.Series) -> pl.Series:
    """Convert European decimal format (1.234,56) to float (1234.56)."""

    def _convert(val: str | None) -> float | None:
        if val is None:
            return None
        v = str(val).strip()
        if "," not in v:
            # No comma present — parse as-is (US format or plain number)
            try:
                return float(v)
            except ValueError:
                return None
        # European format: dots are thousands, comma is decimal
        v = v.replace(".", "").replace(",", ".")
        try:
            return float(v)
        except ValueError:
            return None

    return series.map_elements(_convert, return_dtype=pl.Float64)


@register_transform(
    name="scientific_to_decimal",
    input_types=["string", "numeric"],
    auto_apply=False,
    priority=45,
    mode="series",
)
def scientific_to_decimal(series: pl.Series) -> pl.Series:
    """Convert scientific notation (1.5e3) to decimal (1500.0)."""

    def _convert(val: str | None) -> float | None:
        if val is None:
            return None
        try:
            return float(str(val).strip())
        except ValueError:
            return None

    return series.map_elements(_convert, return_dtype=pl.Float64)

File: packages/goldenflow-js/src/core/transforms/numeric.ts
/**
 * Numeric transforms — ported from goldenflow/transforms/numeric.py
 * Side-effect module: registers 9 numeric transforms on import.
 */

import type { ColumnValue } from "../types.js";
import { registerTransform } from "./registry.js";

// ---------------------------------------------------------------------------
// currency_strip (series, string|numeric, 50)
// ---------------------------------------------------------------------------

function currencyStrip(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    if (typeof v === "number") return v;
    const cleaned = String(v).replace(/[^0-9.\-]/g, "");
    if (cleaned === "" || cleaned === "-") return v;
    const n = Number(cleaned);
    return isNaN(n) ? v : n;
  });
}

registerTransform(
  { name: "currency_strip", inputTypes: ["string", "numeric"], priority: 50, mode: "series" },
  currencyStrip,
);

// ---------------------------------------------------------------------------
// percentage_normalize (series, string|numeric, 50)
// ---------------------------------------------------------------------------

function percentageNormalize(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    if (typeof v === "number") return v / 100;
    const s = String(v).replace(/%/g, "").trim();
    const n = Number(s);
    return isNaN(n) ? v : n / 100;
  });
}

registerTransform(
  { name: "percentage_normalize", inputTypes: ["string", "numeric"], priority: 50, mode: "series" },
  percentageNormalize,
);

// ---------------------------------------------------------------------------
// round (series, numeric, 40, param: n=2)
// ---------------------------------------------------------------------------

function roundTransform(
  values: readonly ColumnValue[],
  n: unknown = 2,
): ColumnValue[] {
  const decimals = typeof n === "number" ? n : Number(n) || 2;
  const factor = Math.pow(10, decimals);

  return values.map((v) => {
    if (v === null) return null;
    const num = typeof v === "number" ? v : Number(v);
    if (isNaN(num)) return v;
    return Math.round(num * factor) / factor;
  });
}

registerTransform(
  { name: "round", inputTypes: ["numeric"], priority: 40, mode: "series" },
  roundTransform,
);

// ---------------------------------------------------------------------------
// clamp (series, numeric, 40, params: min_val=0, max_val=1)
// ---------------------------------------------------------------------------

function clamp(
  values: readonly ColumnValue[],
  minVal: unknown = 0,
  maxVal: unknown = 1,
): ColumnValue[] {
  const lo = typeof minVal === "number" ? minVal : Number(minVal) || 0;
  const hi = typeof maxVal === "number" ? maxVal : Number(maxVal) || 1;

  return values.map((v) => {
    if (v === null) return null;
    const num = typeof v === "number" ? v : Number(v);
    if (isNaN(num)) return v;
    return Math.min(hi, Math.max(lo, num));
  });
}

registerTransform(
  { name: "clamp", inputTypes: ["numeric"], priority: 40, mode: "series" },
  clamp,
);

// ---------------------------------------------------------------------------
// to_integer (series, string|numeric, 45)
// ---------------------------------------------------------------------------

function toInteger(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const num = Number(v);
    if (isNaN(num)) return v;
    return Math.trunc(num);
  });
}

registerTransform(
  { name: "to_integer", inputTypes: ["string", "numeric"], priority: 45, mode: "series" },
  toInteger,
);

// ---------------------------------------------------------------------------
// abs_value (series, numeric, 40)
// ---------------------------------------------------------------------------

function absValue(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const num = typeof v === "number" ? v : Number(v);
    if (isNaN(num)) return v;
    return Math.abs(num);
  });
}

registerTransform(
  { name: "abs_value", inputTypes: ["numeric"], priority: 40, mode: "series" },
  absValue,
);

// ---------------------------------------------------------------------------
// fill_zero (series, numeric, 35) — null -> 0
// ---------------------------------------------------------------------------

function fillZero(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => (v === null ? 0 : v));
}

registerTransform(
  { name: "fill_zero", inputTypes: ["numeric"], priority: 35, mode: "series" },
  fillZero,
);

// ---------------------------------------------------------------------------
// comma_decimal (series, string|numeric, 48) — European "1.234,56" -> 1234.56
// ---------------------------------------------------------------------------

function commaDecimal(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    if (typeof v === "number") return v;
    const s = String(v);
    // European format: dots as thousands separators, comma as decimal
    const converted = s.replace(/\./g, "").replace(",", ".");
    const n = Number(converted);
    return isNaN(n) ? v : n;
  });
}

registerTransform(
  { name: "comma_decimal", inputTypes: ["string", "numeric"], priority: 48, mode: "series" },
  commaDecimal,
);

// ---------------------------------------------------------------------------
// scientific_to_decimal (series, string|numeric, 45)
// ---------------------------------------------------------------------------

function scientificToDecimal(values: readonly ColumnValue[]): ColumnValue[] {
  return values.map((v) => {
    if (v === null) return null;
    const n = Number(v);
    return isNaN(n) ? v : n;
  });
}

registerTransform(
  { name: "scientific_to_decimal", inputTypes: ["string", "numeric"], priority: 45, mode: "series" },
  scientificToDecimal,
);
